### DISCUSSION TITLE
webpack5 with cache is slower than without cache

### DISCUSSION THREAD

**Discussion by lanmch** [webpack/webpack#14282] [General]
Recently, I try to replace babel-loader from esbuild-loader to improve the compile speed of my project. But when I use esbuild-loader, I find the build of project with webpack5, the use "cache: true" is slower than without it. 
I don't know why there is such a phenomenon.
My webpack version is 5.51.0

**Comment by alexander-akait** [comment]
Can you provide steps? cache requires some more time on initial run

**Reply by lanmch** [reply]
provide steps? I get that the caching capability is provided by webpack5 by default. And when I compile my project, I can find that it was compiled twice. Also, the node_modules generated the folder about the cache. But the compilation speed is still very slow

**Reply by lanmch** [reply]
Here is the folder of cacheing.
![image](https://user-images.githubusercontent.com/37408756/134275725-b4107d2b-f0df-4166-9df0-32c442ce38a0.png)

**Reply by alexander-akait** [reply]
Yes, steps, I don't know your configuration

**Reply by alexander-akait** [reply]
webpack doesn't enable filesystem cache by default

**Reply by lanmch** [reply]
> webpack doesn't enable filesystem cache by default

I know it will not enable by default. 
In the previous dialogue, I said that the default probably means that webpack does not require additional configuration such as cache-loader like the old version.
I think i am afraid that i can't show you all config because this config file has too many configuration. How about just the config of cache? Or other property that you want to know.

Here is the config of the cache property.
```json
{
    "type":"filesystem",
    "cacheDirectory":"/Users/lanmch/project/dde/mch-web/.temp/webpackCache",
    "buildDependencies":{
        "configFile":[
            "/Users/lanmch/project/dde/mch-web/dde.config.js"
        ],
        "coraWebpack":[
            "/Users/lanmch/project/dde/mch-web/node_modules/@es/dde-web-build/dist/index.js"
        ],
        "packageJSON":[
            "/Users/lanmch/project/dde/mch-web/package.json"
        ]
    }
}
```

**Reply by alexander-akait** [reply]
Hm, I am afraid something from non official plugins/loaders maybe slow down, maybe you can provide `stats` output or run it using `--profile`/`ProfilePlugin`?

**Reply by lanmch** [reply]
stats like this? It hit the cache
```
assets by path js/*.js 1.16 GiB
  assets by chunk 267 MiB (id hint: commons) 180 assets
  assets by chunk 461 MiB (id hint: vendors) 119 assets
assets by path images/ 45.5 MiB 48 assets
assets by path src/common/assets/favicons/ 95.5 KiB 57 assets
assets by path fonts/ 1.19 MiB
  assets by path fonts/*.ttf 621 KiB 19 assets
  assets by path fonts/*.woff2 283 KiB 15 assets
  assets by path fonts/*.woff 316 KiB 14 assets
assets by path *.html 115 KiB 8 assets
assets by path *.js 5.39 MiB 7 assets
asset 942ded62baea736d15d3.module.wasm 48 KiB [emitted] [immutable] (auxiliary name: cncs_sm4_gcm_wasm)
asset space.ico 4.19 KiB [emitted]
cached modules 29.7 MiB [cached] 23947 modules
runtime modules 279 KiB 128 modules
javascript modules 137 MiB 25885 modules
asset modules 171 KiB (javascript) 1.19 MiB (asset)
  modules by path ./node_modules/katex/dist/ 111 KiB (javascript) 1.18 MiB (asset) 60 modules
  modules by path ./src/pc/ 32.1 KiB (javascript) 10.2 KiB (asset) 5 modules
  modules by path ./node_modules/slick-carousel/slick/ 7.07 KiB 3 modules
  modules by path ./node_modules/@bd/ 11.7 KiB 2 modules
  ./node_modules/video.js/dist/font/VideoJS.eot 9.13 KiB [code generated]
json modules 1.37 MiB
  modules by path ./src/ 454 KiB 34 modules
  modules by path ./node_modules/ 920 KiB 24 modules
  ./package.json 29.6 KiB [code generated]
./node_modules/@byd/wasm/wasm_bg.wasm 115 bytes (javascript) 48 KiB (webassembly) [code generated]
webpack 5.51.1 compiled with 184 warnings in 224820 ms
```

**Reply by alexander-akait** [reply]
> assets by path js/*.js 1.16 GiB

:hear_no_evil: 

What is very big? `1.16 GiB`, it will take a lot of time for writing

**Reply by alexander-akait** [reply]
Problem on initial build only? No warnings/errors?

**Reply by lanmch** [reply]
> > assets by path js/*.js 1.16 GiB
> 
> 🙉
> 
> What is very big? `1.16 GiB`, it will take a lot of time for writing

That might be embarrassing😭. The whole project is close to 20 GiB(include the node_modules), and the src in project is also close to 180MiB. So I think the ‘big’ is the accumulation of all compile products.

**Reply by lanmch** [reply]
> Problem on initial build only? No warnings/errors?

No errors. 184 warning like this log shows.

**Reply by alexander-akait** [reply]
What are warnings? Maybe something break cache, can you show?

**Reply by lanmch** [reply]
warning like this.  I think all warnings are only related to business code and not about webpack compilation.
```
WARNING in ./src/w/components/v-tree/index.ts 2:0-36
export 'UseDrag' (reexported as 'UseDrag') was not found in './useDrag' (possible exports: default)
 @ ./src/w/components/w-tree-v2/wt.tsx 20:0-74 234:41-57 258:17-24
 @ ./src/w/components/w-tree-v2/index.ts 1:0-37 1:0-37
 @ ./src/w/components/w-sidebar/ws.tsx 71:0-39 169:28-36
 @ ./src/w/components/w-sidebar/index.ts 1:0-40 2:0-30 2:0-30 3:15-26
 @ ./src/w/components/w-detail/wd.tsx 60:0-61 768:28-41
 @ ./src/w/components/w-detail/connector.ts 12:0-38 55:60-70
 @ ./src/w/components/w-detail/index.tsx 4:11-6:18
 @ ./src/w/components/w-route-entry/wse.tsx 50:0-57 134:21-31
 @ ./src/w/components/w-route-entry/connector.ts 5:0-46 21:60-74
 @ ./src/w/components/w-route-entry/index.tsx 7:11-9:18
 @ ./src/application/pc/components/route-suite/RouteSuite.tsx 37:0-74 210:75-97
 @ ./src/application/pc/components/route-suite/index.ts 1:0-38 2:15-25
 @ ./src/application/pc/components/route-main/loadables.tsx 83:11-85:44
 @ ./src/application/pc/components/route-main/RouteMain.tsx 9:0-41 154:26-49 158:21-43 189:17-37 205:17-39 257:17-36 262:17-38 267:17-42 274:17-30 279:17-40 290:24-44 300:24-43 312:24-44 320:24-52
 @ ./src/application/pc/components/route-main/index.ts 1:0-36 2:15-24
 @ ./src/pc/app/index.tsx 53:0-58 218:28-34
 @ ./src/pc/index.tsx 113:0-24 450:11-14 454:11-14 458:17-20 459:11-14 464:29-32 467:11-14 471:18-21 476:29-478:3 570:27-30 476:2-478:4 477:11-35
```

**Reply by alexander-akait** [reply]
hm, due that warning, webpack doesn't cache your modules you need to fix it

**Reply by lanmch** [reply]
okay, thanks a lot. I will try it later.

**Reply by alexander-akait** [reply]
Feel free to feedback

**Reply by lanmch** [reply]
I feel so helpless that I have solved 160 warnings(186 total), but the speed does not seem to improve.

**Reply by alexander-akait** [reply]
You can have module on top level with warning and it prevents to cache whole nested modules

**Reply by lanmch** [reply]
sorry, i think i cant understand what do you mean. Can you show me a simple example?

**Reply by alexander-akait** [reply]
Anyway can you debug cache? Using `infrastructureLogger.debug` https://webpack.js.org/configuration/other-options/#debug with `/Cache/`

**Reply by alexander-akait** [reply]
`index.js` can import `./foo.js` and `./foo.js` can import `./bar.js` if `index.js` has warning/errors, we can't cache nested modules

**Reply by lanmch** [reply]
i will try it~
